[Publisher][Agency Settings] Add new agency definition section for court and supervision sectors

nasaownsky opened the pull request:
## Description of the change

Added new agency definition section for court and supervision sectors

## Type of change

> All pull requests must have at least one of the following labels applied (otherwise the PR will fail):

| Label                       	| Description                                                                                               	|
|-----------------------------	|-----------------------------------------------------------------------------------------------------------	|
| Type: Bug                   	| non-breaking change that fixes an issue                                                                   	|
| Type: Feature               	| non-breaking change that adds functionality                                                               	|
| Type: Breaking Change       	| fix or feature that would cause existing functionality to not work as expected                            	|
| Type: Non-breaking refactor 	| change addresses some tech debt item or prepares for a later change, but does not change functionality    	|
| Type: Configuration Change  	| adjusts configuration to achieve some end related to functionality, development, performance, or security 	|
| Type: Dependency Upgrade      | upgrades a project dependency - these changes are not included in release notes                             |

## Related issues

closes #1540 

## Checklists

### Development

**This box MUST be checked by the submitter prior to merging**:
- [ ] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [ ] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

nasaownsky commented:
Hi @nichelle-hall! Please test this feature and tell me your thoughts about my changes!

Also, I have two questions:
1. In figma designs there is [statement](https://www.figma.com/design/mW9Hm6L95zpdWtq0noAdpz?node-id=0-109#971630724):
> If the supervision sector is combined, then we can exclude this entire includes/excludes section from the Agency Settings page.

but in issue [description](https://github.com/Recidiviz/justice-counts/issues/1540) there is another statement:
> If we have the unlikely & unusual case where an agency has both Court, and Supervision sectors, then we should show the all of the relevant sections when they open the modal

Could you please explain to me if my understanding based on my implementation is right or do I need to change something? If the latter, could you maybe tell me in more details about it?

2. There is [comment](https://www.figma.com/design/mW9Hm6L95zpdWtq0noAdpz?node-id=0-109#978377754) in figma designs about free text box, but from endpoint structure description it is not clear where it should be written.

Thank you in advance!

nasaownsky commented:
Oh, okay, I see!
>Do you have other tickets to work on in the meantime?

Sure, I have plenty!

mxosman commented:
Hi @nasaownsky! I'll do a deeper review of the code next week, but just wanted to answer your questions:

> Hi @nichelle-hall! Please test this feature and tell me your thoughts about my changes!
> 
> Also, I have two questions:
> 
> 1. In figma designs there is [statement](https://www.figma.com/design/mW9Hm6L95zpdWtq0noAdpz?node-id=0-109#971630724):
> 
> > If the supervision sector is combined, then we can exclude this entire includes/excludes section from the Agency Settings page.
> 
> but in issue [description](https://github.com/Recidiviz/justice-counts/issues/1540) there is another statement:
> 
> > If we have the unlikely & unusual case where an agency has both Court, and Supervision sectors, then we should show the all of the relevant sections when they open the modal
> 
> Could you please explain to me if my understanding based on my implementation is right or do I need to change something? If the latter, could you maybe tell me in more details about it?

Sorry, this was a bad explanation on my part. I think you implemented it how the ticket was described, but they should be separate sections and separate modals - meaning, if it's a Supervision + Courts agency, they should see two separate sections on the Agency Settings page (one for Supervision Agency Definition and another for Courts Agency Definition). Sorry about the horrible confusion here - I don't know why I combined them as one in my mind when writing that note. Does this clarification make sense?


> 2. There is [comment](https://www.figma.com/design/mW9Hm6L95zpdWtq0noAdpz?node-id=0-109#978377754) in figma designs about free text box, but from endpoint structure description it is not clear where it should be written.

Ah, very good question! For these, you can use the following structure:
```
{ 
  key: "ADDITIONAL_CONTEXT",
  value: "",
}
```

So, we'll have to also allow the setting object to have an optional `value` field and update the `AgencyDefinitionSetting` type. The BE is saving all of this as a JSON blob, so there shouldn't be any issues sending it with this new field as long as the FE is consistent. 

---

One more thing to clarify - the `Supervision Agency Definition` section should not appear if the agency does not have any subpopulations included. 

So it should appear for this configuration (they have two subpopulations included - Parole and Probation):
![Screenshot 2024-11-27 at 4 26 13 PM](https://github.com/user-attachments/assets/18493a3b-4a84-4444-adee-b1aaacbb1c68)

And should NOT appear at all for this configuration:
![Screenshot 2024-11-27 at 4 26 23 PM](https://github.com/user-attachments/assets/8c2d68fd-d30e-4fd3-afe6-53dd1f1979bb)

Let me know if all of this helps clarify or if there are other areas I can help clear up! Thank you for your patience, @nasaownsky.

nasaownsky commented:
Hi @mxosman, thank you for clarification info!

>if it's a Supervision + Courts agency, they should see two separate sections on the Agency Settings page (one for Supervision Agency Definition and another for Courts Agency Definition)

That is hell of a engineering problem! I've tried to came up with the solution for this request, but after several trial and error I realized that this would be not as easy as I initially thought. The main issue is that we have this setting modal as a unified setting `SECTOR_INCLUDES_EXCLUDES`, and all of the actions for opening this modal and so on are happening under this unified name, that represents actual BE endpoint. So even if we split this modal into two separate modals, the edit button would be controlling one `SECTOR_INCLUDES_EXCLUDES` setting, which would be just the latest modal in FE. And even if create some sophisticated separation mechanism, we will just rewrite BE with single set of options (Court or Supervision) every time we will save our Court or Supervision setting in combined agencies. And so while I think it is possible to made some custom solution for this problem, it would be not very elegant and much time consuming. 

So then I came up with another solution: why don't we just display single modal with combined options for combined agencies like this: 

<img width="596" alt="Снимок экрана 2024-12-02 в 16 55 52" src="https://github.com/user-attachments/assets/2cd55401-a358-4e22-a339-8bcc5a024480">

I tested this solution in multiple scenarios and it works just great! I tweaked it so that if it is combined agency and if the agency does not have any subpopulations included -- it will be just `Court Agency Definition` modal, and if it does have some subpopulations -- it will be `Combined Agency Definition` modal with setting options for `courts and pretrial` system and for respective Supervision systems. 

Anyway, please check out my solution and tell me what you think about it! Thank you!

mxosman commented:
> Hi @mxosman, thank you for clarification info!
> 
> > if it's a Supervision + Courts agency, they should see two separate sections on the Agency Settings page (one for Supervision Agency Definition and another for Courts Agency Definition)
> 
> That is hell of an engineering problem! I've tried to came up with the solution for this request, but after several trial and error I realized that this would be not as easy as I initially thought. The main issue is that we have this setting modal as a unified setting `SECTOR_INCLUDES_EXCLUDES`, and all of the actions for opening this modal and so on are happening under this unified name, that represents actual BE endpoint. So even if we split this modal into two separate modals, the edit button would be controlling one `SECTOR_INCLUDES_EXCLUDES` setting, which would be just the latest modal in FE. And even if create some sophisticated separation mechanism, we will just rewrite BE with single set of options (Court or Supervision) every time we will save our Court or Supervision setting in combined agencies. And so while I think it is possible to made some custom solution for this problem, it would be not very elegant and much time consuming.

Oh that's interesting - I actually thought it would be much easier this way. I do see that both settings are unified under the `SECTOR_INCLUDES_EXCLUDES` setting - but further in the `value` property - it's separated by sector. Wouldn't it be possible to achieve the separation and read/write from the `SECTOR_INCLUDES_EXCLUDES` > `value` > `sector` property?

> So then I came up with another solution: why don't we just display single modal with combined options for combined agencies like this:

Ah, that was the goal of the original ticket write up - totally fine this solution if the above is not feasible - especially since this wouldn't be a common case (having both Court + Supervision sectors).

I'm still looking through the code - but, noticed when I was playing with it that:
* it still shows for a Supervision-only agency
* the "Other" text input field does not seem to persist consistently yet (maybe you're still working on this one)

mxosman commented on publisher/src/components/AgencySettings/AgencySettingsDefinition.tsx:
```diff
@@ -0,0 +1,323 @@
+    saveAgencySettings,
+  } = agencyStore;
+
+  const isSupervisionAgency =
```
nit: just for clarity, I might rename this to `isSupervisionAgencyWithEnabledSubpopulations`

mxosman commented on publisher/src/components/AgencySettings/AgencySettingsDefinition.tsx:
```diff
@@ -0,0 +1,323 @@
+    );
+  }, [currentAgencySettings]);
+
+  const isSettingConfigured = agencyDefinitionSetting.length > 0;
```
Will this still be a valid way to check this if the user checks and unchecks everything? I think there will still be fields here if the user unchecks everything.

mxosman replied:
Got it!

mxosman commented on publisher/src/components/AgencySettings/AgencySettingsDefinition.tsx:
```diff
@@ -0,0 +1,323 @@
+  const isSettingConfigured = agencyDefinitionSetting.length > 0;
+
+  useEffect(() => {
+    setCurrentSystems([]);
```
Do we need to reset this every time the effect runs?

nasaownsky replied:
Yeah, unfortunately we need, because when we select agency without applicable sectors we need to reset `currentSystems`, otherwise it will have previous values in it. And I can't do it in reset effect [here](https://github.com/Recidiviz/justice-counts/pull/1594#discussion_r1866377494), because it depends on `currentSystems` and it'll cause infinite loop.

mxosman replied:
Got it - can we add a small comment here to clarify?

mxosman commented on publisher/src/components/AgencySettings/AgencySettingsDefinition.tsx:
```diff
@@ -0,0 +1,323 @@
+
+    if (isCombinedAgency) {
+      setCurrentSystems([
+        ...((currentAgencySystems?.filter((system) =>
+          SupervisionSubsystems.includes(system)
+        ) as AgencySystemKeys[]) || []),
```
We could just have one variable that we set outside of this effect that gives us the list of subsystems and use that variable throughout.

nasaownsky commented:
>it still shows for a Supervision-only agency

You mean it show up when the agency does not have any subpopulations included? It actually working correct, I think you just stumble upon bug with Supervision Populations setting section -- it does not updates UI correctly when switching agencies. Here, I made a loom demonstrating it.

https://www.loom.com/share/4c36b54f83514162834d15519a8fa2d9?sid=5863bcba-4691-465b-875c-02b72ee0d82c

>the "Other" text input field does not seem to persist consistently yet (maybe you're still working on this one)

The field itself showing up for me correctly everytime... Do you mean it is not recording value to BE? That's because I wanted to choose our path forward -- with combined modal or not -- so I won't need to rewrite that functionality if we decide something else!

mxosman commented on publisher/src/components/AgencySettings/AgencySettingsDefinition.tsx:
```diff
@@ -0,0 +1,323 @@
+    isCombinedAgency,
+  ]);
+
+  const defaultSetting = isSettingConfigured
```
nit: maybe `initialSetting` is a more accurate name (to avoid confusion with the actual default settings)?

mxosman commented on publisher/src/components/AgencySettings/AgencySettingsDefinition.tsx:
```diff
@@ -0,0 +1,323 @@
+        ([key, obj]) => {
+          return {
+            key,
+            included: defaultIncluded ?? obj.default,
```
Since we're getting the default settings, why are we also passing in a `defaultIncluded` override param?

mxosman replied:
Ohh I see - thank you for explaining! Hmm - how about `unconfiguredDefault`?

mxosman commented on publisher/src/components/AgencySettings/AgencySettingsDefinition.tsx:
```diff
@@ -0,0 +1,323 @@
+    setUpdatedSetting(defaultSetting);
+    setDescriptionValue("");
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [agencyDefinitionSetting, currentSystems]);
```
Can you clarify the purpose of this effect? We already initialize it with `defaultSetting` and update it via the component, no?

nasaownsky replied:
This is a reset effect used for resetting setting for UI to work properly. The absence of this effect causes visual bug with incorrect working UI when switching agencies like [here](https://github.com/Recidiviz/justice-counts/pull/1594#issuecomment-2512351273)

mxosman replied:
Thanks for clarifying. I'd add a small comment here too for clarity.

mxosman commented on publisher/src/components/AgencySettings/AgencySettingsDefinition.tsx:
```diff
@@ -0,0 +1,323 @@
+                      <CheckboxOptions
+                        options={[
+                          ...Object.entries(AgencyIncludesExcludes[system]).map(
+                            ([key, mapObj]) => {
```
nit: instead of `mapObj`, maybe `option`?

mxosman commented on publisher/src/components/AgencySettings/AgencySettingsDefinition.tsx:
```diff
@@ -0,0 +1,323 @@
+                        onChange={({ key, checked }) => {
+                          setUpdatedSetting((prevSettings) => {
+                            const updates = prevSettings.map((sector) => {
+                              if (sector.sector === system) {
+                                return {
+                                  ...sector,
+                                  settings: sector.settings.map((setting) =>
+                                    setting.key === key
+                                      ? {
+                                          ...setting,
+                                          included: boolToYesNoEnum(!checked),
+                                        }
+                                      : setting
+                                  ),
+                                };
+                              }
+                              return sector;
+                            });
+
+                            return updates;
```
Can we move this to a helper function just for readability that takes in the `prevSettings` and spits out the update?

mxosman commented:
> > it still shows for a Supervision-only agency
> 
> You mean it shows up when the agency does not have any subpopulations included? It actually working correct, I think you just stumble upon bug with Supervision Populations setting section -- it does not updates UI correctly when switching agencies. Here, I made a loom demonstrating it.
> 
> https://www.loom.com/share/4c36b54f83514162834d15519a8fa2d9?sid=5863bcba-4691-465b-875c-02b72ee0d82c

Ahh - interesting... weird that it's not updating when switching over. Thank you for pointing out that it's an unrelated bug. Hopefully it's not too gnarly to fix - thankfully the underlying logic is working and it's just the UI not updating.

Side note: wIthin that same Supervision-only agency in your video - when I turn uncheck all of the subpopulations, it reverts to the Court Agency Definition instead of not showing that section at all.

> > the "Other" text input field does not seem to persist consistently yet (maybe you're still working on this one)
> 
> The field itself showing up for me correctly everytime... Do you mean it is not recording value to BE? That's because I wanted to choose our path forward -- with combined modal or not -- so I won't need to rewrite that functionality if we decide something else!

Sorry for not being clear - yes, I meant it doesn't save the value. Gotcha - please disregard then - my apologies as I thought that part was ready!

nasaownsky commented on publisher/src/components/AgencySettings/AgencySettingsDefinition.tsx:
```diff
@@ -0,0 +1,323 @@
+        >
+          <Modal
+            title={`${agencyTitle} Agency Definition`}
+            description={
```
It seems like the modal can accept `children`, and this used in Admin panel for agency and user provisioning modals, but for this exact purpose it lacks some styling. I think I was looking through code of all settings related components and made it with use of `description` for the sake of uniformity. To sum up: I think we either refactor all settings components to use it as `children` or keep it as `description`.

mxosman replied:
Gotcha! Thank you for clarifying - let's keep it as is!

nasaownsky commented:
@mxosman I resolved visual issues, and answered your question, please check my changes and thank you for your thoughtful comments and questions!

As for textbox implementation you suggested this structure: 

```
{ 
  key: "ADDITIONAL_CONTEXT",
  value: "",
}
```

but our setting is an object that could have multiple sectors, and above structure belongs to only one sector. So, in case of multiple sectors in setting object -- should we write textbox input to each sector or only to first one?

mxosman commented:
> @mxosman I resolved visual issues and added other fixes, please check my changes and thank you for your thoughtful comments and questions!

Taking a look now! Thank _you_ so much for your thoughtful responses and clarifications!

> As for textbox implementation you suggested following structure:
> 
> ```
> { 
>   key: "ADDITIONAL_CONTEXT",
>   value: "",
> }
> ```
> 
> but our setting is an object that could have multiple sectors, and above structure belongs to only one sector. So, in case of multiple sectors in setting object -- should we write textbox input to each sector or only to first one?

Good question! There should be an additional context box for each sector. So if it's a Supervision (w/ enabled subpopulations) agency + Courts agency - there should be two additional context text boxes - one for each sector. If the agency is just a Courts agency (for example), then the additional context belongs just to the courts sector.

Oh, and one quick thing - can we update the placeholder for these context boxes and replace "breakdown" with "agency"? So it reads: 

If the listed categories do not adequately describe your agency, please describe additional data elements included in your agency’s definition.

Current placeholder for reference:
![Uploading Screenshot 2024-12-04 at 10.02.26 AM.png…]()

mxosman commented:
I really appreciate you working through the complexity here - I know these new requests have been super challenging. Just know that all your efforts, creativity and patience are deeply appreciated. 🙏

mxosman commented on publisher/src/components/AgencySettings/AgencySettingsDefinition.tsx:
```diff
@@ -0,0 +1,336 @@
+    removeEditMode();
+  };
+
+  const handleCheckboxChange = (
```
Can you take another look at this? I'm unable to select anything in the checkboxes.

mxosman replied:
It looks like it's just the supervision agency definitions that are affected.

nasaownsky replied:
Fixed that!

mxosman commented on publisher/src/components/AgencySettings/AgencySettingsSupervisions.tsx:
```diff
@@ -72,6 +72,10 @@ export const AgencySettingsSupervisions: React.FC<{
 
+  useEffect(() => {
+    setSupervisionSystemsToSave(currentAgencySystems);
+  }, [currentAgencySystems]);
```
Great call on this!

nasaownsky commented:
@mxosman Thank you for thoughtful comments and observations! This one is really tricky, but I think we are close to an end. I fixed major issue with selection that you pointed out and also added textbox functionality. Please, check my changes and let me know what you think about them. Thank you so much!

mxosman commented on publisher/src/components/AgencySettings/AgencySettings.styles.tsx:
```diff
@@ -444,6 +444,13 @@ export const SupervisionSystemRow = styled(AgencySettingsInfoRow)<{
   border: none;
 `;
 
+// Agency Defifnition
```
tiniest nit: typo in Definition

mxosman commented on publisher/src/components/AgencySettings/AgencySettingsDefinition.tsx:
```diff
@@ -215,6 +220,28 @@ const AgencySettingsDefinition: React.FC<{
 
+  const handleDescriptionChange = (system: AgencySystemKeys, value: string) => {
+    setUpdatedSetting((prevSettings) => {
+      return prevSettings.map((sector) => {
```
tiny nit: `setting` instead of `sector` for the mapped param.

nasaownsky replied:
Oh, this actually resolved in [this](https://github.com/Recidiviz/justice-counts/pull/1594/commits/36f5dc29fac587b642199fb76efceddc03b2ae1c) commit

mxosman replied:
Ah, sorry - I was looking at the wrong commit! Thank you so much!

mxosman reviewed (approved):
Ah, this looks and feels great! Thank you for all of your work on this, @nasaownsky! This was all tricky indeed - hopefully we'll get a break soon from all the trickiness in these requests. Left a few tiny nits and question about what ended up fixing the checkbox issue - but otherwise looks great and is ready to move on to staging - the end for this is here! Thanks again!

nasaownsky commented on publisher/src/components/AgencySettings/AgencySettingsDefinition.tsx:
```diff
@@ -152,15 +150,22 @@ const AgencySettingsDefinition: React.FC<{
+
+  const initialSetting = [
+    ...agencyDefinitionSetting,
+    ...getDefaultSetting(unconfiguredSystems, IncludesExcludesEnum.NO),
```
Yeah, the problem was that once systems were written to BE, there was no way for unconfigured (or unwritten) systems to get into `initialSetting`

mxosman replied:
Gotcha! Great catch and sleuthing!

nasaownsky commented:
Yay! Thank you so much for all your help on this @mxosman!
